CHANGELOG.md

```diff
@@ -1,3 +1,514 @@
+  # The name can be any valid K8s string for name. This example uses
+  # the following convention: cstor-vol-<flattened-from-to-versions>
+  name: cstor-vol-11101120
+
+  # VERIFY the value of namespace is same as the namespace where openebs components
+  # are installed. You can verify using the command:
+  # `kubectl get pods -n <openebs-namespace> -l openebs.io/component-name=maya-apiserver`
+  # The above command should return status of the openebs-apiserver.
+  namespace: openebs
+
+spec:
+  backoffLimit: 4
+  template:
+    spec:
+      # VERIFY the value of serviceAccountName is pointing to service account
+      # created within openebs namespace. Use the non-default account.
+      # by running `kubectl get sa -n <openebs-namespace>`
+      serviceAccountName: openebs-operator
+      containers:
+      - name:  upgrade
+        args:
+        - "cstor-volume"
+
+        # --from-version is the current version of the volume
+        - "--from-version=1.11.0"
+
+        # --to-version is the version desired upgrade version
+        - "--to-version=1.12.0"
+
+        # Bulk upgrade is supported from 1.9
+        # To make use of it, please provide the list of cstor volumes
+        # as mentioned below
+        - "pvc-3415af20-db82-42cf-99e0-5d0f2809c657"
+        - "pvc-c3d0b587-5da9-457b-9d0e-23331ade7f3d"
+        - "pvc-e115f3f9-1666-4680-a932-d05bfd049087"
+        # For upgrades older than 1.9.0, use
+        # '--pv-name=<pv_name> format as
+        # below commented line
+        # - "--pv-name=pvc-c630f6d5-afd2-11e9-8e79-42010a800065"
+
+        #Following are optional parameters
+        #Log Level
+        - "--v=4"
+        #DO NOT CHANGE BELOW PARAMETERS
+        env:
+        - name: OPENEBS_NAMESPACE
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace
+        tty: true
+
+        # the image version should be same as the --to-version mentioned above
+        # in the args of the job
+        image: quay.io/openebs/m-upgrade:1.12.0
+        imagePullPolicy: Always
+      restartPolicy: OnFailure
+---
+
+```
+
+Apply the Job manifest using `kubectl`. Check the logs of the pod started by the Job:
+
+```console
+$ kubectl get logs -n openebs cstor-vol-1001120-8b2h9
+..
+..
+..
+I0903 12:41:41.984635       1 cstor_volume_upgrade.go:609] Upgrade Successful for cstor volume pvc-e115f3f9-1666-4680-a932-d05bfd049087
+I0903 12:41:41.994013       1 cstor_volume.go:119] Successfully upgraded cstorVolume{pvc-e115f3f9-1666-4680-a932-d05bfd049087} from 1.11.0 to 1.12.0
+```
+
+Verify that all the volumes are updated to the latest version by running the following command:
+
+```console
+$ kubectl get cstorvolume -A -o jsonpath='{.items[*].versionDetails.status.current}'
+1.12.0 1.12.0 1.12.0
+```
+
+#### Upgrade other components
+
+Other components are safe to upgrade by running the following command:
+
+```
+lokoctl component apply <component name>
```

Comment:
After upgrading MetalLB and storage workloads, other components can be updated automatically, there is no need to update one by one.
```suggestion
lokoctl cluster apply
```

Reply:
I am wary of doing it. I would prefer it to be done one at a time because if a complex set of events break some component you have no way of knowing what broke it? So I would ask the users to do one component at a time, verify things are fine then move to next.

Reply:
Do you have a particular breakage in mind? I think this is pretty bad UX if the user needs to type all remaining components manually. `lokoctl component apply` should work, and if it doesn't we should fix it so I'd accept this change.

Reply:
> Do you have a particular breakage in mind?

No.

> I think this is pretty bad UX if the user needs to type all remaining components manually. `lokoctl component apply` should work, and if it doesn't we should fix it so I'd accept this change.

I agree. But also it is safer. I was testing the networking components like metallb, contour, cert-manager and I upgraded them once using command `lokoctl component apply metallb contour cert-manager` and things started failing now I wasn't sure why. Eventually I figured it is metallb only when I did an individual update. When metallb failed, nothing failed visually (like pods were fine everywhere) looking at things, but when I looked closely at the logs of metallb I figured it is failing. So such unknown scenarios I want to avoid, hence the upgrade to do it individually. If `lokoctl` would have co-ordinated upgrades making sure things are fine at every step and then moved ahead I would not hesitate in recommending it then. But now it is better to be safer.

Reply:
Fair.